File: bench/bench.py
"""Python binding side of the benchmark; mirrors bench/bench.cpp exactly.

Run both through bench/run.sh, which compiles the C++ baseline against the
same PCL install the extension was built against and diffs the results.

Usage: python bench/bench.py [n_points] [repeats] [pcd_path]
"""

from __future__ import annotations

import os
import sys
import time

import numpy as np

import pcl


def report(name: str, best: float) -> None:
    print(f"{name}\t{best:.9f}")


def bench(fn, repeats: int) -> float:
    """Minimum wall-clock time over `repeats` runs (matches bench.cpp)."""
    best = float("inf")
    for _ in range(repeats):
        t0 = time.perf_counter()
        fn()
        best = min(best, time.perf_counter() - t0)
    return best


def main(argv: list[str]) -> int:
    n = int(argv[1]) if len(argv) > 1 else 1000000
    repeats = int(argv[2]) if len(argv) > 2 else 5
    path = argv[3] if len(argv) > 3 else "/tmp/bench_py.pcd"

    # Deterministic source data, identical formula to bench.cpp.
    idx = np.arange(n, dtype=np.float32)
    src = np.empty((n, 3), dtype=np.float32)
    src[:, 0] = idx * 0.001
    src[:, 1] = idx * 0.002
    src[:, 2] = idx * 0.003

    cloud = pcl.PointCloud()

    def do_from_array():
        c = pcl.PointCloud()
        c.from_array(src)
        return c

    report("from_array", bench(do_from_array, repeats))
    cloud = do_from_array()

    report("to_array", bench(cloud.to_array, repeats))
    report("save_pcd_binary",
           bench(lambda: pcl.save(cloud, path, binary=True), repeats))
    report("load_pcd_binary", bench(lambda: pcl.load(path), repeats))

    # Per-point access: n round trips through the wrapper. This is the
    # pathological case a binding cannot make fast; it exists to quantify
    # how much the bulk path above is worth.
    def per_point_sum():
        sx = sy = sz = 0.0
        for i in range(n):
            x, y, z = cloud[i]
            sx += x
            sy += y
            sz += z
        return sx + sy + sz

    report("per_point_sum", bench(per_point_sum, repeats))

    # Centroid computed the idiomatic way: one bulk transfer, then numpy.
    def centroid():
        arr = cloud.to_array()
        return arr.mean(axis=0)

    report("centroid", bench(centroid, repeats))

    print(f"points {n}  repeats {repeats}", file=sys.stderr)
    if os.path.exists(path):
        os.remove(path)
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: bench/bench.cpp
// Native C++ baseline for bench/bench.py.
//
// Both programs run the SAME workloads on the same point counts so the
// difference is exactly the cost of going through the Python/Cython
// wrapper. Timings are the minimum of R repetitions (least noisy
// estimator for short CPU-bound runs); output is `name<TAB>seconds` so
// bench/compare.py can diff the two runs mechanically.
#include <pcl/io/pcd_io.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <string>
#include <vector>

using Clock = std::chrono::steady_clock;
using Cloud = pcl::PointCloud<pcl::PointXYZ>;

static double elapsed(Clock::time_point a, Clock::time_point b) {
    return std::chrono::duration<double>(b - a).count();
}

static void report(const char* name, double best) {
    std::printf("%s\t%.9f\n", name, best);
}

int main(int argc, char** argv) {
    const std::size_t n = (argc > 1) ? std::strtoul(argv[1], nullptr, 10) : 1000000;
    const int repeats = (argc > 2) ? std::atoi(argv[2]) : 5;
    const std::string path = (argc > 3) ? argv[3] : "/tmp/bench_cpp.pcd";

    // Deterministic source data, identical formula to bench.py.
    std::vector<float> src(n * 3);
    for (std::size_t i = 0; i < n; ++i) {
        src[i * 3 + 0] = static_cast<float>(i) * 0.001f;
        src[i * 3 + 1] = static_cast<float>(i) * 0.002f;
        src[i * 3 + 2] = static_cast<float>(i) * 0.003f;
    }

    // --- from_array: build a cloud from a flat float buffer ---------------
    Cloud::Ptr cloud(new Cloud);
    double best = 1e300;
    for (int r = 0; r < repeats; ++r) {
        Cloud::Ptr c(new Cloud);
        auto t0 = Clock::now();
        c->width = static_cast<std::uint32_t>(n);
        c->height = 1;
        c->is_dense = true;
        c->points.resize(n);
        for (std::size_t i = 0; i < n; ++i) {
            c->points[i].x = src[i * 3 + 0];
            c->points[i].y = src[i * 3 + 1];
            c->points[i].z = src[i * 3 + 2];
        }
        auto t1 = Clock::now();
        best = std::min(best, elapsed(t0, t1));
        cloud = c;
    }
    report("from_array", best);

    // --- to_array: copy the cloud back out into a flat float buffer -------
    best = 1e300;
    for (int r = 0; r < repeats; ++r) {
        std::vector<float> out(n * 3);
        auto t0 = Clock::now();
        for (std::size_t i = 0; i < n; ++i) {
            out[i * 3 + 0] = cloud->points[i].x;
            out[i * 3 + 1] = cloud->points[i].y;
            out[i * 3 + 2] = cloud->points[i].z;
        }
        auto t1 = Clock::now();
        best = std::min(best, elapsed(t0, t1));
    }
    report("to_array", best);

    // --- save (binary PCD) -----------------------------------------------
    best = 1e300;
    for (int r = 0; r < repeats; ++r) {
        auto t0 = Clock::now();
        pcl::io::savePCDFile(path, *cloud, true);
        auto t1 = Clock::now();
        best = std::min(best, elapsed(t0, t1));
    }
    report("save_pcd_binary", best);

    // --- load (binary PCD) ------------------------------------------------
    best = 1e300;
    for (int r = 0; r < repeats; ++r) {
        Cloud::Ptr c(new Cloud);
        auto t0 = Clock::now();
        pcl::io::loadPCDFile(path, *c);
        auto t1 = Clock::now();
        best = std::min(best, elapsed(t0, t1));
    }
    report("load_pcd_binary", best);

    // --- per-point access: sum every coordinate one point at a time -------
    // The pathological case for a binding: n round trips through the
    // wrapper instead of one bulk transfer.
    best = 1e300;
    double checksum = 0.0;
    for (int r = 0; r < repeats; ++r) {
        double sx = 0.0, sy = 0.0, sz = 0.0;
        auto t0 = Clock::now();
        for (std::size_t i = 0; i < n; ++i) {
            const pcl::PointXYZ& p = cloud->points[i];
            sx += p.x;
            sy += p.y;
            sz += p.z;
        }
        auto t1 = Clock::now();
        best = std::min(best, elapsed(t0, t1));
        checksum = sx + sy + sz;
    }
    report("per_point_sum", best);

    // --- centroid: the same result computed the idiomatic way -------------
    best = 1e300;
    for (int r = 0; r < repeats; ++r) {
        auto t0 = Clock::now();
        double sx = 0.0, sy = 0.0, sz = 0.0;
        for (std::size_t i = 0; i < n; ++i) {
            sx += cloud->points[i].x;
            sy += cloud->points[i].y;
            sz += cloud->points[i].z;
        }
        volatile double cx = sx / static_cast<double>(n);
        volatile double cy = sy / static_cast<double>(n);
        volatile double cz = sz / static_cast<double>(n);
        (void)cx; (void)cy; (void)cz;
        auto t1 = Clock::now();
        best = std::min(best, elapsed(t0, t1));
    }
    report("centroid", best);

    std::fprintf(stderr, "checksum %.6f  points %zu  repeats %d\n",
                 checksum, n, repeats);
    std::remove(path.c_str());
    return 0;
}
